pyrules/rules.py

```diff
@@ -1,93 +1,161 @@
     """
-    condition = lambda self, context: False
-    action = lambda self, context: None 
+    def __init__(self, condition=None, action=None):
+        self._condition = condition
+        self._action = action
+
+    def condition(self, context):
+        """
+        Condition for executing this rule.
+
+        Override in subclasses if necessary. Should return boolean value that
+        determines if rule is used.
+        """
+        return self._condition(self, context)
+
+    def action(self, context):
+        """
+        Action for executing this rule.
+
+        Override in subclasses if necessary. Should return dictionary with
+        results that will be added to context.
+        """
+        return self._action(self, context)
+
     def should_trigger(self, context):
         return self.condition(context)
+
     def perform(self, context):
         result = self.action(context) 
         context.update(result)
         return result
-    
-class TableRuleset(Rule):
+
+
+class TableRule(Rule):
     """
-    a table ruleset created from a list of dict objects of the following format
+    A table rule is created from a list of dict objects of the following format:
     
     [ 
       {
-        'if' : ['condition1', ...],
+        'if' : {'logic': '1 | 2', 'conditions': ['foo', {'bar': 10}]},
         'then' : ['action1', ...],
         'target' : ['target1', ...]
       }, 
       ...
     ]
     
-    Each rule is only executed if all conditions are met. In conditions and actions, use context. 
-    to reference variables. targets implicitly reference context. (target 'xy' means 'context.xy').
-    The result of the nth 'then' action is stored in the nth context.variable as defined in target.
+    Each rule is only executed if all conditions are met. In actions, use
+    'context.' to reference variables. Targets and conditions implicitly reference
+    'context.' (target 'xy' means 'context.xy'). Logic can be omitted, which
+    would imply "&" operation for all conditions. Condition can be a dictionary or
+    a single value, so 'value' is equivalent to {'value': True}
+
+    The result of the nth 'then' action is stored in the nth 'context.variable'
+    as defined in target.
     """
-    def __init__(self, rules, translations=None):
+    def __init__(self, rules, name=None):
         self.rules = rules or {}
-        if translations:
-            translator = Translator(translations)
-            for rule in self.rules:
-                for key in rule.keys():
-                    rule[key] = [translator.replace(item) for item in rule[key]]
-    def should_trigger(self, context):
-        return True
+        if name:
+            self.name = name
+        self._current_ruleid = None
+
     def perform(self, context):
         count = 0
         for rule in self.rules:
-            if all([eval(condition, context.as_dict) for condition in rule['if']]):
+            evaluator = LogicEvaluator(
```

Comment:
> So we sort of add compile time checks, while currently it only has runtime checks.

Yes. Let's give these "bits" names, please. Can we agree on the format being 

* yaml input: `var op value` 
* kwarg: `var__op=value` 

Is there some other meaning that I didn't catch?

> But that's NOT what parsing does

Depends on the point of view, I guess. IMHO ExpressionHandler.evaluate currently parses and executes at the same time, which it shouldn't for the reason above.

To me, parsing should check is that the condition is in valid syntax, and that the operator is known. Right now that would include checking that the format is as expected `[bit__]*bit` or better the `var op value` syntax. Obviously it can't check if the variable is valid because it doesn't know the context yet - that's a runtime matter.

Also I don't think we should impose the internal `var__op=value` format on the user. That is a matter of passing input to `C`, and it is great format there, but it is not a good format the human readable/writeable YAML.

> your example has a mistake (or typo) - what you ask for is required finding errors like "second gt: 5", without the ":" just running from_yaml 

I'm not sure that's the case. This is perfectly valid YAML:

```
condition: 
- if: second gt 5
```

but the ExpressionHandler will still throw an error.